alf/algorithms/algorithm.py

```diff
@@ -1070,6 +1075,7 @@ def calc_loss(self, experience, train_info):
             " calc_loss() to generate LossInfo from train_info")
         return train_info
 
+    @common.mark_training
```

Comment:
As long as the main algorithm and sub-algorithm don't run in parallel, this should be fine because we also want the sub-algorithm to know whether it's training or not. In my use case, the main algorithm and sub-algorithm share an observation normalizer which only updates the stats in the main algorithm rollout_step.

Reply:
It's possible that a sub-algorithm calls train_from_replay_buffer() during a train_from_unroll() from the higher level algorithm. The training flag will be changed by the sub-algorithm.

Reply:
It's better to pass a flag to transform_timestep() when it's called.

Reply:
> It's possible that a sub-algorithm calls train_from_replay_buffer() during a train_from_unroll() from the higher level algorithm. The training flag will be changed by the sub-algorithm.

All training iters happen sequentially, I think. So there is no nested train iters. The reason why I didn't use a flag is because the observation_transformer can be any function (e.g., image_scale_transformer, cast_transformer). So the API is hard to be unified. Right now for any transformer we directly call it on observation as the sole argument.  (Unless we also set some global training flag instead of passing the flag to transformer.)

Reply:
Or maybe maintain a stack of training flags?

Reply:
If obs_normalizer is the only place training flag is needed, then since it's not really useful, maybe we just remove from this PR?

Reply:
Added just for potential future use. Sooner or later we will face this issue.